### ISSUE TITLE
Reducers are definitely not pure

### ISSUE THREAD

**Issue by markerikson** [cliffhall/react-chat-client#6]
Redux reducers are supposed to be "pure functions".  That means they are only `(state, action) => newState`, with no side effects - no async calls, no sockets, no accessing the rest of the store directly, no global variables, or anything else.

These lines are definitely _not_ supposed to be in a reducer:

https://github.com/cliffhall/react-chat-client/blob/master/src/store/message/reducer/index.js#L87-L93

They should go into a middleware instead.  See the [Redux FAQ entry on where websockets should live](https://redux.js.org/faq/codestructure#where-should-websockets-and-other-persistent-connections-live) for examples.

**Comment by cliffhall** [comment]
Excellent pint Mark. I definitely would not have put that there today. I will revisit this to relocate that call.

**Comment by markerikson** [comment]
Thanks.  I ran across this because a learner in the Reactiflux chat channels had been looking at it and was very confused by it.  Since it looks like this is related to a tutorial, I'd appreciate it if you could fix that up so it's no longer an issue :)

**Comment by cliffhall** [comment]
Fixed with PR #7

**Comment by cliffhall** [comment]
@markerikson A lot of changes happened because I also wanted to remove use of dispatch from components and instead create dispatcher functions in mapDispatchToProps. 

To your point, the one reducer function that was not pure and was interacting with the socket to send a message has been cleaned up. Also a couple of components (SendButton and ConnectButton) were interacting with the socket instead of sending actions, and Client was creating the Socket and its listeners and passing the socket instance down the component hierarchy. All that has been moved to the new middleware: https://github.com/cliffhall/react-chat-client/blob/master/src/store/socket/middleware/index.js

I've also updated the accompanying tutorial here: https://itnext.io/building-a-react-based-chat-client-with-redux-816b47cb8c74
